lib/common/xml.c

```diff
@@ -1371,33 +1371,32 @@ mark_child_moved(xmlNode *old_child, xmlNode *new_parent, xmlNode *new_child,
 static void
 mark_xml_changes(xmlNode *old_xml, xmlNode *new_xml, bool check_top)
```

Comment:
This function was definitely the hardest one to get right (I hope), so it's probably also the hardest one to review over the course of several commits. I broke it into pieces as best I could.

The `xml_accept_changes()`/`pcmk__xml_commit_changes()` part was also kind of tough to reason about, since existing code was doing some things that seemed unnecessary.